File: Week Assignment 7 using Python/count_substrings_k_distinct.py
def count_at_most_k(s, k):
    freq = {}
    left = 0
    res = 0
    for right in range(len(s)):
        freq[s[right]] = freq.get(s[right], 0) + 1
        while len(freq) > k:
            freq[s[left]] -= 1
            if freq[s[left]] == 0:
                del freq[s[left]]
            left += 1
        res += right - left + 1
    return res

def count_exactly_k(s, k):
    return count_at_most_k(s, k) - count_at_most_k(s, k - 1)

s = input().strip()
k = int(input())
print(count_exactly_k(s, k))

File: Week Assignment 7 using C/count_substrings_k_distinct.c
#include <stdio.h>
#include <string.h>
#include <stdbool.h>
#define MAX 256

int countAtMostK(char *s, int k) {
    int freq[MAX] = {0}, distinct = 0, res = 0;
    int left = 0;
    for (int right = 0; s[right]; right++) {
        if (freq[s[right]] == 0) distinct++;
        freq[s[right]]++;
        while (distinct > k) {
            freq[s[left]]--;
            if (freq[s[left]] == 0) distinct--;
            left++;
        }
        res += right - left + 1;
    }
    return res;
}
int countExactlyK(char *s, int k) {
    return countAtMostK(s, k) - countAtMostK(s, k - 1);
}
int main() {
    char s[1000]; int k;
    scanf("%s %d", s, &k);
    printf("%d", countExactlyK(s, k));
    return 0;
}